pennylane/data/base/mapper.py

```diff
@@ -0,0 +1,141 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains a class for mapping HDF5 groups to Dataset Attributes, and a mixin
+class that provides the mapper class."""
+
+
+import typing
+from collections.abc import MutableMapping
+from types import MappingProxyType
+from typing import Any, Dict, Optional, Type
+
+from pennylane.data.base.attribute import (
+    AttributeInfo,
+    DatasetAttribute,
+    get_attribute_type,
+    match_obj_type,
+)
+from pennylane.data.base.hdf5 import HDF5Group
+
+
+class AttributeTypeMapper(MutableMapping):
+    """
+    This class performs the mapping between the objects contained
+    in a HDF5 group and Dataset attributes.
+    """
+
+    bind: HDF5Group
+    _cache: Dict[str, DatasetAttribute]
+
+    def __init__(self, bind: HDF5Group) -> None:
+        self._cache = {}
+        self.bind = bind
+
+    def __getitem__(self, key: str) -> DatasetAttribute:
+        if key in self._cache:
+            return self._cache[key]
+
+        h5_obj = self.bind[key]
+
+        attr_type = get_attribute_type(h5_obj)
+        attr = attr_type(bind=h5_obj)
+        self._cache[key] = attr
+
+        return attr
+
+    @property
+    def info(self) -> AttributeInfo:
+        """Return ``AttributeInfo`` for ``self.bind``."""
+        return AttributeInfo(self.bind.attrs)
+
+    def set_item(
+        self,
+        key: str,
+        value: Any,
+        info: Optional[AttributeInfo],
+        require_type: Optional[Type[DatasetAttribute]] = None,
+    ) -> None:
+        """Creates or replaces attribute ``key`` with ``value``, optionally
+        including ``info``.
+
+        Args:
+            key: Name of attribute in HDF5 group
+            value: Attribute value, either a compatible object or an already
+                initialized ``DatasetAttribute``.
+            info: Extra info to attach to attribute
+            require_type: Force the ``value`` to be serialized as this type.
+                If ``value`` is an ``DatasetAttribute``, it must be an instance of ``require_type``.
+                Otherwise, ``value`` must be serializable by ``require_type``.
+        """
+        if isinstance(value, DatasetAttribute):
+            if require_type and not isinstance(value, require_type):
+                raise TypeError(key, require_type, type(value))
```

Comment:
can this be more informative? something like `f"expected {key} to be of type {require_type}, but got {type(value)} instead."`